includes/wikia/api/SendGridPostBackApiController.class.php

```diff
@@ -1,91 +1,115 @@
 <?php
 /**
- * API which SendGrid will post data to when an email is opened, clicked, bounced back, unsubscribed, marked as spam, etc..
+ * API which SendGrid will post data to when an email is opened, clicked, etc..
  *
- * @author Piotr Molski ( moli@wikia-inc.com )
+ * For more information see:
+ *
+ *   https://sendgrid.com/docs/API_Reference/Webhooks/event.html
  */
 
-class SendGridPostbackController extends WikiaApiController {
+use Wikia\Logger\WikiaLogger;
 
-	public function safe_get($arr, $key, $default = null) {
-		return isset($arr[$key]) ? $arr[$key] : $default;
-	}
+class SendGridPostbackController extends WikiaApiController {
 
+	/**
+	 * Requests from SendGrid are POSTed via this controller.  The POST data contains a JSON
+	 * encoded associative array with the following keys:
+	 *
+	 * - email
+	 * - wikia-email-id
+	 * - wikia-email-city-id
+	 * - wikia-db
+	 * - event
+	 * - url
+	 * - status
+	 * - reason
+	 *
+	 * NOTE: The 'response' key set before returning is for debugging purposes only.  SendGrid
+	 * only requires that we send back a 200 response code.
+	 *
+	 */
 	public function log() {
-		wfProfileIn( __METHOD__ );
-
-		// Sendgrid uses raw JSON POST format, this is how PHP ingests it
-		$data = file_get_contents("php://input");
-		$events = json_decode($data, true);  // true = return value is array
-
-		if (empty($events)) {
-			\Wikia\Logger\WikiaLogger::instance()->debug( __CLASS__ . ": No data to process");
-			wfProfileOut( __METHOD__ );
+		if ( !$this->wg->Request->wasPosted() ) {
+			// Log the token-validation problem.
+			WikiaLogger::instance()->debug( __CLASS__ . ': Request method must be POST' );
+			$this->response->setVal( 'response', 'Request method must be POST' );
 			return;
 		}
 
-		if (!is_array($events)) {
-			\Wikia\Logger\WikiaLogger::instance()->debug( __CLASS__ . ": Supplied data is not an array", [
-				'data_first_100_chars' => is_string($data) ? substr($data,0,100) : '[type: ' . gettype($data) . ']'
-			]);
-			wfProfileOut( __METHOD__ );
+		$events = $this->getEvents();
+		if ( empty( $events ) ) {
+			$this->response->setVal( 'response', 'Unable to find any events' );
 			return;
 		}
 
-		foreach ($events as $event) {
```

Comment:
The only thing this code did was move certain data from one array to another.  There's no need to do this when passing to WikiaLogger and we can pull off the individual values when we need them.  I gave a link to the docs above and listed the keys we use for future reference.